session_test.go

```diff
@@ -663,6 +734,57 @@ func TestKeepAlive(t *testing.T) {
 	}
 }
 
+func TestKeepAlive_Timeout(t *testing.T) {
+	conn1, conn2 := testConn()
+
+	clientConf := testConf()
+	clientConf.ConnectionWriteTimeout = time.Hour // We're testing keep alives, not connection writes
+	clientConf.EnableKeepAlive = false            // Just test one direction, so it's deterministic who hangs up on whom
+	client, _ := Client(conn1, clientConf)
+	defer client.Close()
+
+	server, _ := Server(conn2, testConf())
+	defer server.Close()
+
+	clientLogs := captureLogs(client)
+	serverLogs := captureLogs(server)
+
+	errCh := make(chan error, 1)
+	go func() {
+		_, err := server.Accept() // Wait until server closes
+		errCh <- err
+	}()
+
+	// Prevent the client from responding
+	clientConn := client.conn.(*pipeConn)
+	clientConn.writeBlocker.Lock()
+
+	select {
+	case err := <-errCh:
+		if err != ErrKeepAliveTimeout {
+			t.Fatalf("unexpected error: %v", err)
+		}
+	case <-time.After(1 * time.Second):
+		t.Fatalf("timeout waiting for timeout")
+	}
+
+	if !client.IsClosed() {
```

Comment:
Running locally it looks like the test is flaky some times right here.

```
workpad:yamux james$ untilfail go test ./...
ok      _/Users/james/projects/rnapier/yamux    12.165s
ok      _/Users/james/projects/rnapier/yamux    11.947s
2015/09/10 10:47:53 [ERR] yamux: keepalive failed: session shutdown
2015/09/10 10:47:53 [ERR] yamux: keepalive failed: session shutdown
--- FAIL: TestKeepAlive_Timeout-4 (0.35s)
        session_test.go:772: client should have closed
2015/09/10 10:48:03 [WARN] yamux: failed to send ping reply: connection write timeout
FAIL
FAIL    _/Users/james/projects/rnapier/yamux    11.800s
```

You could probably remove the client-side checks since all the action is on the server side for this one, or add a small delay for the client to finish the close. I'm puzzled who is emitting `[WARN] yamux: failed to send ping reply: connection write timeout` though, that must be the server but it should be able to do the write.